### PR TITLE
Fixes testUpdateBranchGrayRulesWithUpdateOnce

### PR DESCRIPTION
**Pull request by ZhewenFu**
## What's the purpose of this PR

Fix test `testUpdateBranchGrayRulesWithUpdateOnce`

## Which issue(s) this PR fixes:

Fixes the issue that causes flaky tests. The current test may lead to inconsistent ordering of stringified JSON objects and the test may fail sometimes. To fix the test and make it consistent, I select all the elements in the list and see whether the release history contains those elements. In this case, the ordering of the elements doesn't matter and won't cause failure.

## Brief changelog

Added containRule function.

### CODE REVIEW

**Comment by ZhewenFu** [comment]
Sorry I closed my previous PR by accident. The reason I didn't use [GrayReleaseRuleItemTransformer#batchTransformFromJSON](https://github.com/apolloconfig/apollo/blob/301d66dcc28b2d698eefed3ecc9aa3273afb9ddf/apollo-common/src/main/java/com/ctrip/framework/apollo/common/utils/GrayReleaseRuleItemTransformer.java#L35) is once we transform it into object, the object's `toString()` method still requires us to perform string manipulation.

**Comment by zhewenfufufu** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by zhewenfufufu** [comment]
recheck

File: apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java [hunk 1]
```diff
@@ -64,25 +64,32 @@ public void testUpdateBranchGrayRulesWithUpdateOnce() {
 
     namespaceBranchService.updateBranchGrayRules(testApp, testCluster, testNamespace, testBranchName, rule);
 
-    GrayReleaseRule
-        activeRule =
-        namespaceBranchService.findBranchGrayRules(testApp, testCluster, testNamespace, testBranchName);
+    GrayReleaseRule activeRule = namespaceBranchService.findBranchGrayRules(testApp, testCluster, testNamespace,
```
Comment (nobodyiam) [inline apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java:67]: It's better not to involve unnecessary changes, as they would make it hard to distinguish what's actually changed.

File: apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java [hunk 2]
```diff
@@ -64,25 +64,32 @@ public void testUpdateBranchGrayRulesWithUpdateOnce() {
...
+        testNamespace, pageable);
 
     ReleaseHistory releaseHistory = releaseHistories.getContent().get(0);
 
     Assert.assertEquals(1, releaseHistories.getTotalElements());
     Assert.assertEquals(ReleaseOperation.APPLY_GRAY_RULES, releaseHistory.getOperation());
     Assert.assertEquals(0, releaseHistory.getReleaseId());
     Assert.assertEquals(0, releaseHistory.getPreviousReleaseId());
-    Assert.assertTrue(releaseHistory.getOperationContext().contains(rule.getRules()));
+    Assert.assertTrue(containRule(releaseHistory.getOperationContext(), rule.getRules()));
```
Comment (nobodyiam) [inline apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java:84]: I understand the original `contains` check is not a good one, however, I would like to know the background behind this change. Is this test failed in some situations?
BTW, I would suggest parsing the string to object for comparison instead of simply comparing the string, as there is not much difference from the original check.

File: apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java (hunk 1 above)
Comment (ZhewenFu) [reply apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java]: Thanks for pointing out! It was caused by autoformatting and now I have restored them to the original version.

File: apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java (hunk 2 above)
Comment (ZhewenFu) [reply apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java]: > I understand the original `contains` check is not a good one, however, I would like to know the background behind this change. Is this test failed in some situations? BTW, I would suggest parsing the string to object for comparison instead of simply comparing the string, as there is not much difference from the original check.

* Is this test failed in some situations?

  Yes, the test failed in certain cases.  I detected that the test results are flaky using the NonDex tool for catching non-deterministic/flaky tests (https://github.com/TestingResearchIllinois/NonDex).

  You may reproduce the nondeterministic behavior using the following steps:
  Clone the repo and cd into it.
  `mvn install -am -pl RosettaCore -DskipTests`
  `mvn -pl RosettaCore test -Dtest=com.hubspot.rosetta.RosettaBinderTest#itHandlesAnnotatedOptionalGenericSetterSerialization - Confirm that the test passes as is.`
  `mvn -pl RosettaCore edu.illinois:nondex-maven-plugin:1.1.2:nondex -DnondexMode=ONE -Dtest=com.hubspot.rosetta.RosettaBinderTest#itBindsStoredAsJsonBeanCorrectly` - This will run the NonDex tool on the test. 

* Parsing the string to object for comparison instead of simply comparing the string

  Thanks for pointing this out! I have modified the code and currently, I am using [GrayReleaseRuleItemTransformer#batchTransformFromJSON](https://github.com/apolloconfig/apollo/blob/301d66dcc28b2d698eefed3ecc9aa3273afb9ddf/apollo-common/src/main/java/com/ctrip/framework/apollo/common/utils/GrayReleaseRuleItemTransformer.java#L35) to transfer the string to object and compare the element within the set. Let me know if you have any suggestions on this :)

**Comment by nobodyiam** [comment]
The unit test failed, which seems related to the line of code

```java
Set<GrayReleaseRuleItemDTO> contextRules = GrayReleaseRuleItemTransformer.batchTransformFromJSON(context.substring(context.indexOf("["),context.lastIndexOf("]")+1));
```

**Comment by ZhewenFu** [comment]
> The unit test failed, which seems related to the line of code
> 
> ```java
> Set<GrayReleaseRuleItemDTO> contextRules = GrayReleaseRuleItemTransformer.batchTransformFromJSON(context.substring(context.indexOf("["),context.lastIndexOf("]")+1));
> ```

Yeah that seems to be where the problem lies. I have changed it to transfer the context to a map struct to avoid hardcoding. Now it should work fine.

File: apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java [hunk 3]
```diff
@@ -115,10 +122,24 @@ public void testUpdateBranchGrayRulesWithUpdateTwice() {
...
-    Assert.assertFalse(firstReleaseHistory.getOperationContext().contains(secondRule.getRules()));
-    Assert.assertTrue(secondReleaseHistory.getOperationContext().contains(firstRule.getRules()));
-    Assert.assertTrue(secondReleaseHistory.getOperationContext().contains(secondRule.getRules()));
+    Assert.assertTrue(containRule(firstReleaseHistory.getOperationContext(), firstRule.getRules()));
+    Assert.assertFalse(containRule(firstReleaseHistory.getOperationContext(), secondRule.getRules()));
+    Assert.assertTrue(containRule(secondReleaseHistory.getOperationContext(), firstRule.getRules()));
+    Assert.assertTrue(containRule(secondReleaseHistory.getOperationContext(), secondRule.getRules()));
+  }
+
+  private boolean containRule(String context, String rule) {
```
Comment (nobodyiam) [inline apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java:131]: How about we refactor this method a little bit to actually check whether the context contains some rule? It's not necessary for current test cases, but we could use that in the future.

```java
  private boolean containRules(String context, String rules) {
    Type grayReleaseRuleItemsType = new TypeToken<Map<String, Set<GrayReleaseRuleItemDTO>>>() {
    }.getType();
    Map<String, Set<GrayReleaseRuleItemDTO>> contextRulesMap = new Gson().fromJson(context, grayReleaseRuleItemsType);
    Set<GrayReleaseRuleItemDTO> ruleSet = GrayReleaseRuleItemTransformer.batchTransformFromJSON(rules);

    for (GrayReleaseRuleItemDTO rule : ruleSet) {
      boolean found = false;
      for (Set<GrayReleaseRuleItemDTO> contextRules : contextRulesMap.values()) {
        for (GrayReleaseRuleItemDTO contextRule : contextRules) {
          if (contextRule.toString().equals(rule.toString())) {
            found = true;
            break;
          }
        }
      }
      if (!found) {
        return false;
      }
    }

    return true;
  }
```

Comment (ZhewenFu) [reply apollo-biz/src/test/java/com/ctrip/framework/apollo/biz/service/NamespaceBranchServiceTest.java]: Yes, it looks excellent! This version does make more sense and makes the tests more consistent. Also thanks for fixing some style issues here :)

**Review by nobodyiam — APPROVED** [review]
LGTM
